### PR TITLE
Randomises placement of gulag and mining outpost within lavaland

### PR DESCRIPTION
**Pull request by S34NW**
## What Does This PR Do
This PR transforms the gulag and mining outpost into ruin templates, and forces them to spawn every time.
This has the effect of randomising the gulag and mining station locations every shift.

- No fauna can spawn within a certain radius to prevent early miner deaths by snu-snu.
- Both gulag and mining outpost now have permanent GPS beacons
- Gulag is still separated from the rest of lavaland by a lava river mostly 2 tiles wide (but 1 at some places)
- The areas that were always lava to the south and west of lavaland have been filled in, allowing all but the top section of the map to be used for ruin generation.
- Apart from specified turfs, the "blank" parts of the templates are filled in with whatever turf lies underneath due to map generation.
- A minor remap of the gulag happens here to more easily permit sec access to the ore redemption machine, hopefully making gulag more viable for use as sec aren't guaranteed to get mugged on entry.

I am unsure how this will play out, I feel a period of testing would be required to see if it works for us or not.

Note: These _cannot_ be randomly placed right up against the map border, there is a minimum of 25 tiles from the edge at all time due to how `/datum/map_template/ruin/proc/try_to_place(z,allowed_areas)` works

## Why It's Good For The Game
Mining is a bit "paint by numbers" currently. We have seen how diversity improves the experience since lava rivers have been turned on, but this PR takes it a step further.
While it does not make it extremely more challenging, it allows for a bit of variation that can hopefully play into interesting situations.
You now must rely on your GPS if you want to get home, no more heading south until you hit lava, then heading west.

## Changelog
:cl:
tweak: Gulag and Mining Outpost are now randomly placed within Lavaland, don't forget your GPS!
tweak: Permanent GPS beacon added to Lavaland Gulag.
/:cl:

### CODE REVIEW

File: code/datums/ruins/lavaland.dm [hunk 1]
```diff
@@ -114,13 +130,14 @@
 	cost = 0
 	allow_duplicates = FALSE
 
+/** Yeets this to stop it spawning
```
Comment (AffectedArc07) [inline code/datums/ruins/lavaland.dm:133]: Uncomment this

File: config/example/config.toml [hunk 2]
```diff
@@ -681,21 +681,21 @@ active_lava_ruins = [
...
 	# Re-removed 7/24/2020, I ran a month-long test of hierophant Jun-Jul and the staff was still too strong. https://github.com/ParadiseSS13/Paradise/pull/13542
-	"_maps/map_files/RandomRuins/LavaRuins/lavaland_surface_hierophant.dmm",
+	#"_maps/map_files/RandomRuins/LavaRuins/lavaland_surface_hierophant.dmm",
 	"_maps/map_files/RandomRuins/LavaRuins/lavaland_surface_swarmer_crash.dmm",
 
 	##MISC
 	"_maps/map_files/RandomRuins/LavaRuins/lavaland_surface_animal_hospital.dmm",
 	# Cube blacklisted on production because it contains a wishgranter that gives you hijack on use
-	"_maps/map_files/RandomRuins/LavaRuins/lavaland_surface_cube.dmm",
+	#"_maps/map_files/RandomRuins/LavaRuins/lavaland_surface_cube.dmm",
```
Comment (AffectedArc07) [inline config/example/config.toml:698]: Undo these config changes

**Comment by Miraviel** [comment]
This looks super amazing!

One question, can it spawn close to the north side of the map where the Necropolis is? I see the mining sttion cannot spawn within 25 tiles from the edge, but the necropolis ruin is more than two screens tall, so it may be able to spawn close to the lava river separating it from the rest of the asteroid?

I am only worried about new players seeing a lava river, going "oh, nice bridge", crossing it, "oh, nice door", opening the gate, then releasing the big skull thingy on the station.

**Comment by Fox-McCloud** [comment]
Certain fixtures on lavaland should not be randomized---the base and the gulag fall solidly within this category as they're the entry point to the entire map and also play a role in prisoner containment, which is pretty important.

I don't think this is a good idea *at all*, to say the least.

**Comment by S34NW** [comment]
> This looks super amazing!
> 
> One question, can it spawn close to the north side of the map where the Necropolis is? I see the mining station cannot spawn within 25 tiles from the edge, but the necropolis ruin is more than two screens tall, so it may be able to spawn close to the lava river separating it from the rest of the asteroid?
> 
> I am only worried about new players seeing a lava river, going "oh, nice bridge", crossing it, "oh, nice door", opening the gate, then releasing the big skull thingy on the station.

25 tiles from the edge of the spawnable zone is more correct. The spawnable zone stops before the north lava river so the ruins are shunted down appropriately

**Comment by S34NW** [comment]
> Certain fixtures on lavaland should not be randomized---the base and the gulag fall solidly within this category as they're the entry point to the entire map and also play a role in prisoner containment, which is pretty important.
> 
> I don't think this is a good idea *at all*, to say the least.

I don't see how them being involved in these matters makes any difference? They both have GPS beacons so players who are diligent will be able to find them.
The gulag, if anything, becomes more viable with this as prisoners can't just jump the one tile gap to the mining station infirmary. People have been reluctant to use the gulag because of how easy it is for someone to escape.

**Comment by warriorstar-orion** [comment]
I would love to at least see this TMed, I think this idea has potential but I'd like to see how it actually works out in practice with ruin generation.

Why is the lava from around the map borders removed? I see you say "The areas that were always lava to the south and west of lavaland have been filled in, allowing all but the top section of the map to be used for ruin generation." but it's not clear why that means the lava border needs to be removed. Is it just to provide even more clearance and space for ruins?

**Comment by S34NW** [comment]
> I would love to at least see this TMed, I think this idea has potential but I'd like to see how it actually works out in practice with ruin generation.


Yes, I will be requesting TM to assess viability soon.

> 
> Why is the lava from around the map borders removed? I see you say "The areas that were always lava to the south and west of lavaland have been filled in, allowing all but the top section of the map to be used for ruin generation." but it's not clear why that means the lava border needs to be removed. Is it just to provide even more clearance and space for ruins?

Good question, the lava was removed for that exact reason, to allow more area for ruins. While there is a minimum 23 tile mandated gap from the edge of the area where ruins can spawn to the start of a ruin, this unfortunately applies from the _area_ not the map boundary. Extending the main lavaland area to cover the pre-generated terrain will not work either as it would mean nests/fauna/megas spawning over that lava area. It also would mess up maze/river generation.
So my best solution was to just fill it all in and let the river and maze generator handle things solo.

**Comment by LiberatedWaveMan** [comment]
> Certain fixtures on lavaland should not be randomized---the base and the gulag fall solidly within this category as they're the entry point to the entire map and also play a role in prisoner containment, which is pretty important.
> 
> I don't think this is a good idea _at all_, to say the least.

Even if this is true, at the very least it's in good interest for us to move gulag _away_ from the mining outpost by some extent, so that antags can't just go "hehe i'll bring ice and burn ointment with me!" and make the easiest escape known to man.

**Comment by AffectedArc07** [comment]
This has 2 maint objections, a head objection, and no one in favour from either party.

Concerns include:
- Shuttle code
- "These 2 landmarks should NEVER be randomised"

**Pull request closed without merging**